clients/src/main/java/org/apache/kafka/clients/producer/internals/ProducerBatch.java

```diff
@@ -291,6 +303,45 @@ private void completeFutureAndFireCallbacks(long baseOffset, long logAppendTime,
         return batches;
     }
 
+    public boolean completed() {
+        return produceFuture.completed();
+    }
+
+    public void await() throws InterruptedException {
+        produceFuture.await();
+        waitForPossibleSplittedBatches();
+        clearChildrenProducerBatch();
+    }
+
+    private void waitForPossibleSplittedBatches() throws InterruptedException {
+        for (ProducerBatch child : getChildrenProducerBatch()) {
+            child.produceFuture.await();
+            child.waitForPossibleSplittedBatches();
+            child.clearChildrenProducerBatch();
+        }
+    }
+
+    private void addChildrenProducerBatch(final ProducerBatch batch) {
+        if (childrenProducerBatch == null) {
+            childrenProducerBatch = new ArrayList<>();
+        }
+
+        childrenProducerBatch.add(batch);
+    }
+
+    private List<ProducerBatch> getChildrenProducerBatch() {
```

Comment:
If we could get rid of null check, `addChildrenProducerBatch` and `getChildrenProducerBatch` could be removed as well.